Use cgo to compile protobuf and rocksdb.

petermattis opened the pull request:
This makes cockroach almost go-gettable (only the vendoring of etcd stands in the way).

Intead of compiling protobuf and rocksdb explicitly, we use "go build" to build them by creating wrapper go packages for the C++ sources. See https://github.com/cockroachdb/c-protobuf and https://github.com/cockroachdb/c-rocksdb. This was initially an experiment to see if it was possible. Turns out it is, though there are caveats.

I'm not at all set on this approach and it definitely has some downsides (e.g. duplicating the build systems of these dependent libraries). But the build work is straightforward and we should be bumping the version of rocksdb infrequently and protobuf very rarely. Currently only tested on Mac OS X and Linux and will likely require tweaks for other systems.

petermattis committed:
Use cgo to compile protobuf and rocksdb.

This makes cockroach almost go-gettable (only the vendoring of etcd
stands in the way).

Intead of compiling protobuf and rocksdb explicitly, we use "go build"
to build them by creating wrapper go packages for the C++ sources.

bdarnell commented on proto/Makefile:
```diff
@@ -16,7 +16,7 @@
 # Author: Andrew Bonventre (andybons@gmail.com)
 
-PROTOC      := ../_vendor/usr/bin/protoc
+PROTOC      := protoc
```
This means that bootstrap.sh won't install protoc any more, and users will be responsible for installing it onto their PATH separately with their preferred packaging tools (if they ever change .proto files). Seems reasonable, but will need to be called out in documentation.

petermattis replied:
It was straightforward to build protoc using the same cgo approach. See `github.com/cockroachdb/c-protobuf/cmd/protoc`.

bdarnell replied:
Ah, cool. We should refer to the protoc built in this way with `../../../../../bin/` as we do for protoc-gen-gogo below. That way we don't require `$GOPATH/bin` to be in `$PATH`.

petermattis replied:
Done.

bdarnell commented:
Neat idea. The build work doesn't seem *too* onerous; it's similar to what it would take to use these projects with `gyp` (in fact, I could imagine a `gyp` "backend" that would spit out symlink farms and `cgo_flags.go` files like you have in the other repos).

My main concern with this approach is that there is little value in being *almost* compatible with `go get`. There doesn't seem to be much practical benefit to this change unless we go the rest of the way and change the way we do raft as well (and avoid introducing similar build special-cases in the future). I do think that's a good idea, but it's tricky. We have not yet reached the point of a long-term stable API for raft, so even if we resolved the import path issues we would still need version pinning for this dependency (but at least that's a better-understood problem so we could be compatible with `godep` or some similar tool)

bdarnell commented on storage/engine/cgo_flags.go:
```diff
@@ -0,0 +1,15 @@
+
+// #cgo CXXFLAGS: -std=c++11
+// #cgo CPPFLAGS: -I ../../../c-protobuf/internal/src -I ../../../c-rocksdb/internal/include
+// #cgo darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup
```
LDFLAGS are propagated from one package to its dependents; do we need to put this here (as opposed to leaving it in the c-rocksdb and c-protobuf repos)?

Is this really a darwin vs linux issue, or are we using OS as a proxy for GCC vs clang-based toolchains? The limited flexibility of go build flags might be a problem here if we want to support more platforms.

petermattis replied:
The problem is here, not in the upstream packages. The issue is with how "go build" links the compiled objects into a single object and does not tolerate undefined symbols. Frankly, it seems like a bug in cgo. Previously this wasn't a problem because the `-lrocksdb`, `-lprotobuf`, etc flags are part of the link line and those libraries contain the undefined symbols. 

You are correct that this is not darwin vs !darwin but clang vs gcc. I'm not aware of a build tag which allows you to distinguish clang vs gcc. I'll add a comment explaining this.

spencerkimball commented on storage/engine/cgo_flags.go:
```diff
@@ -0,0 +1,15 @@
+package engine
+
+import (
+	// Link against the protobuf and rocksdb libaries. This is
```
s/This is/This is done/

petermattis replied:
Done.

spencerkimball commented:
LGTM

spencerkimball commented:
LGTM

petermattis commented:
@bdarnell Agreed that this isn't terribly useful unless we also fix the vendoring of raft. I actually thought this was the trickier part of achieving go-gettability. Raft might not be ready for fixing at this moment, but it seems fixable in the medium term.

bdarnell commented:
Yeah, this is the trickier part in the sense that it required more research and experimentation. The problems of version pinning are better understood, it's just that the solutions are all imperfect and whatever we choose will have a greater day-to-day impact on our workflow than this change will.

petermattis commented:
Yes, this change will have little impact on day-to-day workflow. How frequently are you making changes to etcd/raft? From a quick scan of the logs it looks like historically it has been once a week or so. Let's chat about this more tomorrow.

bdarnell commented:
Changes to raft are not that common, especially now that all the changes I was making back in December are done. But when we have changed things, those changes have to propagate in lockstep across all the cockroach developers, which we've done so far with submodules.

petermattis committed:
Specify multiple packages on the "get get" command lines.

This allows the gets to run in parallel. Add a "godeps" make target for
easy updating of the list of dependencies.

petermattis commented:
Rather than using submodules to pin versions, could we ask developers to re-run `bootstrap.sh`. I'm also wondering if we should try to bake the git SHAs of dependencies into the build. It would be straightforward to do.

bdarnell commented:
Re-running `bootstrap.sh` works for developers, but it means we're not `go get` compatible (or rather we don't get the benefit of pinned versions when `go get` is used). `go get` will fetch submodules for us and checkout the right version but for anything that's not a submodule it just gets the head of the default branch.

petermattis committed:
Explicitly use the protoc in ${GOPATH}/bin.

petermattis commented:
Yes, `go get` will use the head version of dependencies. I don't think this is a big problem. We can educate users to build release versions of cockroach using "make build" (or use a standard container we provide). And developers can use `bootstrap.sh` to make sure they are on the right versions of dependencies. Let me know if this sounds too onerous. Otherwise I'd like to merge this change and then see about replacing the vendoring of coreos/etcd.

bdarnell commented:
Yeah, I think that's fine (it's strictly better than what we're doing now in terms of compatibility with the standard tools). But just for the record, it's not only release versions - the concern is not just that `go get` will build combinations that haven't been tested together; we frequently make changes in raft that don't even build without synchronized changes in the client code. 

Go ahead and merge this and I'll look into preparing multinode for a merge upstream so we can replace vendoring of etcd. Maybe also break raft out into a separate repo while we're at it.
